Add FoldM1

Topsii opened the pull request:
I do not export the constructor `FoldM1'`. Instead I export the pattern synonym `FoldM1`, which has a Monad constraint. This way we retain the option to later change the datatype definition to 
```haskell
data FoldM1 m a b = forall x. FoldM1 (a -> m x) (x -> a -> m x) (x -> m b)
```
Note that currently `hoists` or the Functor instance can only be implemented with `FoldM1'` because they do not require `Monad m`.

I also changed the handler type signatures from `Handler a b` to `Handler s a` to match the `s t a b` naming scheme for optics.

Topsii committed:
Add FoldM1

Topsii committed:
Rename FromMaybe's type parameter: b -> a

Topsii committed:
Rename type variables of optics according to s t a b scheme

Gabriella439 commented on src/Control/Foldl/NonEmpty.hs:
```diff
@@ -375,16 +601,77 @@ minimumBy cmp = Fold1 (\begin -> Fold min' begin id)
 {-# INLINABLE minimumBy #-}
 
--- | Upgrade a fold to accept the 'Fold1' type
+-- | Upgrade a "non-empty" fold to accept the 'Fold1' type
```
I think you want to escape quotes, otherwise haddock tries to interpret them as  a module reference.  Same comment for the other quotes in the documentation:

```suggestion
-- | Upgrade a \"non-empty\" fold to accept the 'Fold1' type
```

Gabriella439 reviewed (approved):
Just one comment, otherwise this looks great!

Topsii committed:
Support Haddock's shaded inline code with rounded corners

With background shading the rounded corners between two immediately
consecutive inline code spans would be visible. Merges these
inline code spans into a single one to avoid this.
